File: scala/prime.scala
class Sieve(var prime : Int) {
  var next : Sieve = _

  def test(n : Int) : Unit = {
    if (prime * (n / prime) == n)
      ()
    else if (next == null) {
      println(n)
      next = new Sieve(n)
    } else next.test(n)
  }
}

object Driver {
  def main(args: Array[String]): Unit = {
    val s: Sieve = new Sieve(2)
    var i : Int = 2

    println("2")

    while (i < 10) {
      i = i + 1
      s.test(i)
    }
  }
}

File: java/Prime.java
public class Prime {

  static class Sieve {
    private Sieve next;
    private int prime;

    Sieve(int prime) {
      this.prime = prime;
    }

    void test(int n) {
      if (this.prime * (n / this.prime) == n) {

      } else if (this.next == null) {
        System.out.println(n);
        this.next = new Sieve(n);
      } else {
        this.next.test(n);
      }
    }
  }

  public static void main(String[] args) {

    Sieve s = new Sieve(2);
    int i = 2;

    System.out.println("2");

    while (i < 13) {
      i += 1;
      s.test(i);
    }
  }
}
